stan/math/rev/fun/LDLT_factor.hpp

```diff
@@ -3,140 +3,93 @@
- * d2 = log_determinant_ldlt(ldlt_A2);
- * ~~~
- *
- * @tparam R number of rows, can be Eigen::Dynamic
- * @tparam C number of columns, can be Eigen::Dynamic
+ * @tparam T type of elements in the matrix
  */
-template <int R, int C>
-class LDLT_factor<var, R, C> {
+template <typename T>
+class LDLT_factor<T, true> {
+ private:
+  using ldlt_type = Eigen::LDLT<T>;
+  ldlt_type* ldlt_ptr_;
+
  public:
+  template <typename S, require_eigen_t<S>* = nullptr>
+  explicit LDLT_factor(const S& matrix)
+      : ldlt_ptr_(make_chainable_ptr(matrix.ldlt())) {}
+
   /**
-   * Default constructor.  The caller *MUST* call compute() after this.  Any
-   * calls which use the LDLT_factor without calling compute() run the risk
-   * of crashing Stan from within Eigen.
+   * Return a const reference to the underlying matrix
    */
-  LDLT_factor() : alloc_(new LDLT_alloc<R, C>()) {}
-
-  explicit LDLT_factor(const Eigen::Matrix<var, R, C> &A)
-      : alloc_(new LDLT_alloc<R, C>()) {
-    compute(A);
-  }
+  const auto& matrix() const { return ldlt_ptr_->matrixLDLT(); }
 
   /**
-   * Use the LDLT_factor object to factorize a new matrix.  After calling
-   * this function, the user should call success() to check that the
-   * factorization was successful. If the factorization is not successful,
-   * the LDLT_factor is not valid and other functions should not be used.
-   *
-   * @param A A symmetric positive definite matrix to factorize
+   * Return a const reference to the LDLT factor of the matrix
    */
-  inline void compute(const Eigen::Matrix<var, R, C> &A) {
-    check_square("comute", "A", A);
-    alloc_->compute(A);
-  }
+  const auto& ldlt() const { return *ldlt_ptr_; }
+};
 
 /**
- * Compute the actual numerical result of inv(A)*b.  Note that this isn't
- * meant to handle any of the autodiff.  This is a convenience function
- * for the actual implementations in mdivide_left_ldlt.
- *
- * Precondition: success() must return true. If success() returns false,
- *    this function runs the risk of crashing Stan from within Eigen.
- *
- * @param b The right handside.  Note that this is templated such that
- * Eigen's expression-templating magic can work properly here.
+ * An LDLT_factor of an `Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>`
+ * with `alloc_in_arena = True` holds a copy of the input matrix and the LDLT
+ * of its values, with all member variable allocations are done in the arena.
  */
-#if EIGEN_VERSION_AT_LEAST(3, 3, 0)
-  template <typename Rhs>
-  inline const Eigen::Solve<Eigen::LDLT<Eigen::Matrix<double, R, C> >, Rhs>
-  solve(const Eigen::MatrixBase<Rhs> &b) const {
-    return alloc_->ldlt_.solve(b);
-  }
-#else
-  template <typename Rhs>
-  inline const Eigen::internal::solve_retval<
-      Eigen::LDLT<Eigen::Matrix<double, R, C> >, Rhs>
-  solve(const Eigen::MatrixBase<Rhs> &b) const {
-    return alloc_->ldlt_.solve(b);
-  }
-#endif
+template <>
+class LDLT_factor<Eigen::Matrix<var, Eigen::Dynamic, Eigen::Dynamic>, true> {
```

Comment:
I think it would be nicer / safer if this used SFINAE so if a user passed `const Eigen::Matrix<var, -1, -1>&` here the require would still catch it

Reply:
also idt this scheme would work for `arena_matrix<Eigen::Matrix<var, -1, -1>>`

Reply:
Fair. My intention was to always build these through `stan::math::make_ldlt_factor` which can kinda hide this stuff.

Reply:
I did SFINAE on the class, but I kept the factory function around cuz I like it too.